src/SignalR/clients/ts/signalr/src/HttpConnection.ts

```diff
@@ -376,6 +413,83 @@ export class HttpConnection implements IConnection {
     }
 
+    private async reconnect(error?: Error) {
+        try {
+            await this.startPromise;
+        } catch (e) {
+            // This exception is returned to the user as a rejected Promise from the start method
+            // or was observed by the last reconnect loop.
+        }
+
+        const reconnectPolicy = this.options.reconnectPolicy as IReconnectPolicy;
+        let previousRetryCount = 0;
+        let nextRetryDelay = reconnectPolicy.nextRetryDelayInMilliseconds(previousRetryCount++, 0);
+
+        if (nextRetryDelay === null) {
+            this.logger.log(LogLevel.Information, "Connection not reconnecting because of the IReconnectPolicy.");
+            this.stopConnection(error);
+            return;
+        }
+
+        if (!this.changeState(ConnectionState.Connected, ConnectionState.Reconnecting)) {
+            return;
+        }
+
+        if (error) {
+            this.logger.log(LogLevel.Information, `Connection reconnecting because of error '${error}'.`);
+        } else {
+            this.logger.log(LogLevel.Information, "Connection reconnecting.");
+        }
+
+        if (this.onreconnecting) {
+            this.onreconnecting(error);
+
+            // Exit early if the onreconnecting callback called connection.stop().
+            if (this.connectionState !== ConnectionState.Reconnecting) {
+                return;
+            }
+        }
+
+        const startTime = Date.now();
+
+        while (nextRetryDelay != null) {
+            this.logger.log(LogLevel.Information, `The next reconnect attempt will start in ${nextRetryDelay} ms.`);
+            await new Promise((resolve) => setTimeout(resolve, nextRetryDelay as number));
+
+            if (this.connectionState !== ConnectionState.Reconnecting) {
+                return;
+            }
+
+            try {
+                await this.startInternal(this.transferFormat as TransferFormat);
+
+                // The TypeScript compiler thinks that this.connectionState is always Reconnecting here meaning this condition is always false.
+                // The TypeScript compiler is wrong.
+                if ((this.connectionState as any) === ConnectionState.Connected) {
+                    this.logger.log(LogLevel.Information, "Connection reconnected.");
+
+                    if (this.onreconnected) {
+                        this.onreconnected();
+                    }
+                }
+
+                return;
+            } catch (e) {
+                this.logger.log(LogLevel.Information, `Reconnect attempt failed because of error '${e}'.`);
+
+                if (this.connectionState !== ConnectionState.Reconnecting) {
+                    return;
+                }
+            }
+
+            nextRetryDelay = reconnectPolicy.nextRetryDelayInMilliseconds(previousRetryCount++, Date.now() - startTime);
+        }
+
+        this.logger.log(LogLevel.Information, "Reconnect retry attempts have been exhausted. Connection disconnecting.");
```

Comment:
IMO it would be useful info to have how many reconnect attempts before failure in the log message.